base_layer/core/src/chain_storage/blockchain_database.rs

```diff
@@ -0,0 +1,259 @@
+        let metadata = Self::read_metadata(&db)?;
+        Ok(BlockChainDatabase {
+            metadata: Arc::new(RwLock::new(metadata)),
+            db: Arc::new(db),
+        })
+    }
+
+    /// If a call to any metadata function fails, you can try and force a re-sync with this function. If the RWLock
+    /// is poisoned because a write attempt failed, this function will replace the old lock with a new one with data
+    /// freshly read from the underlying database. If this still fails, there's probably something badly wrong and
+    /// the thread will panic.
+    pub fn try_recover_metadata(&mut self) -> bool {
+        if !self.metadata.is_poisoned() {
+            // metadata is fine. Nothing to do here
+            return false;
+        }
+        match BlockChainDatabase::read_metadata(self.db.as_ref()) {
+            Ok(data) => {
+                self.metadata = Arc::new(RwLock::new(data));
+                true
+            },
+            Err(e) => {
+                error!(
+                    target: LOG_TARGET,
+                    "Could not read metadata from database. {}. We're going to panic here. Perhaps restarting will \
+                     fix things",
+                    e.to_string()
+                );
+                panic!("Blockchain metadata is compromised and the recovery attempt failed.")
+            },
+        }
+    }
+
+    fn get_metadata(&self) -> Result<RwLockReadGuard<ChainMetadata>, ChainStorageError> {
+        self.metadata.read().map_err(|e| {
+            error!(
+                target: LOG_TARGET,
+                "An attempt to get sa read lock on the blockchain metadata failed. {}",
+                e.to_string()
+            );
+            ChainStorageError::AccessError("Read lock on blockchain metadata failed".into())
+        })
+    }
+
+    /// Returns the height of the current longest chain. This method will only fail if there's a fairly serious
+    /// synchronisation problem on the database. You can try calling [BlockchainDatabase::try_recover_metadata] in
+    /// that case to re-sync the metadata; or else just exit the program.
+    pub fn get_height(&self) -> Result<u64, ChainStorageError> {
+        let metadata = self.get_metadata()?;
+        Ok(metadata.height_of_longest_chain)
+    }
+
+    /// Returns the total accumulated work/difficulty of the longest chain.
+    ///
+    /// This method will only fail if there's a fairly serious synchronisation problem on the database. You can try
+    /// calling [BlockchainDatabase::try_recover_metadata] in that case to re-sync the metadata; or else
+    /// just exit the program.
+    pub fn get_total_work(&self) -> Result<Difficulty, ChainStorageError> {
+        let metadata = self.get_metadata()?;
+        Ok(metadata.greatest_accumulated_work.into())
+    }
+
+    /// Returns the transaction kernel with the given hash.
+    pub fn fetch_kernel(&self, hash: HashOutput) -> Result<Option<TransactionKernel>, ChainStorageError> {
```

Comment:
Suggestion: It might be helpful to have these return `TransactionKernel` instead of `Option<TransactionKernel>` and return an error on NotFound, then create an impl that adds `.optional()` to `Result<T, ChainStorageError>` that converts it to an option like we did for BigNeon:  https://github.com/big-neon/bn-api/blob/a61327c2a5e19a8b8181dc35a41412bf0adca617/db/src/utils/errors.rs#L371

Reply:
Yeah, my philosophy here was the `Err` represents an error on the backend (i.e. server error) only. Asking for non-existent records is more a client error. 
I think I'll leave it as is for now and see what the ergonomics of this API are like.